### PR TITLE
I have made some changes in filefield_sources module . Please review and merge

### PR DESCRIPTION
**Pull request by aritnath1990**
I have made some changes regarding token_replace function and attach plugin. Please review the code.

### CODE REVIEW

**Comment by tienvx** [comment]
Thanks for your contribution and sorry for late response. I'll review the merge request soon.

File: filefield_sources.module [hunk 1]
```diff
@@ -188,7 +190,7 @@ function filefield_sources_field_process(&$element, FormStateInterface $form_sta
   if (!isset($enabled_sources['upload'])) {
     foreach (array('upload_button', 'upload') as $field) {
       if (isset($element[$field])) {
-        $element[$field]['#access'] = FALSE;
+        $element[$field]['#access'] = false;
```
Comment (tienvx) [inline filefield_sources.module]: I think constants should be all-uppercase, see [Coding Standards](https://www.drupal.org/coding-standards#naming). I'm on 8.0.x and I think it's still using that standard.

**Comment by aritnath1990** [comment]
I have changed false to False

File: src/Plugin/FilefieldSource/Attach.php [hunk 2]
```diff
@@ -53,7 +57,7 @@ public static function value(array &$element, &$input, FormStateInterface $form_
 
       // Clean up the file name extensions and transliterate.
       $original_filepath = $filepath;
-      $new_filepath = filefield_sources_clean_filename($filepath, $instance->settings['file_extensions']);
+      $new_filepath = filefield_sources_clean_filename($filepath, $instance->getSetting['file_extensions']);
```
Comment (tienvx) [inline src/Plugin/FilefieldSource/Attach.php]: I think brackets are wrong here. getSetting is a method, not a property, so it should be:
```
$instance->getSetting('file_extensions');
```
Please update Remote plugin too. There are 2 places that have this code:
```
$field->settings['file_extensions']
````

**Comment by tienvx** [comment]
I was busy these days, so I read the changes but unfortunately I did not have a chance to run the test cases, so I did not review your merge request completely. So sorry for that. I'll back in a couple of days.

**Comment by aritnath1990** [comment]
No Problem. I have fixed the issue you mentioned in attach and remote plugin. Though I did not get a chance to test the remote plugin. Please have a look when you have some time.

Also Can you please review filefield_sources_flysystem module? I have created this module to integrate flysystem with filefield_sources module.
https://www.drupal.org/node/2604294
git clone --branch 8.x-1.x http://git.drupal.org/sandbox/arit_nath/2604164.git filefield_sources_flysystem

File: filefield_sources.module [hunk 3]
```diff
@@ -169,11 +169,13 @@ function filefield_sources_field_process(&$element, FormStateInterface $form_sta
     elseif (isset($source['class'])) {
       $callback = array($source['class'], 'process');
       if (is_callable($callback)) {
-        $element = call_user_func_array($callback, array(
-          $element,
-          $form_state,
-          $complete_form,
-        ));
+        $element = call_user_func_array(
```
Comment (tienvx) [inline filefield_sources.module]: Please revert this change.

File: filefield_sources.module [hunk 4]
```diff
@@ -283,11 +285,13 @@ function filefield_sources_field_validate(&$element, FormStateInterface $form_st
 
     $callback = array($source['class'], 'validate');
     if (is_callable($callback)) {
-      call_user_func_array($callback, array(
-        $element,
-        $form_state,
-        $complete_form,
-      ));
+      call_user_func_array(
```
Comment (tienvx) [inline filefield_sources.module]: Please revert this change.

File: filefield_sources.module [hunk 5]
```diff
@@ -298,9 +302,9 @@ function filefield_sources_field_validate(&$element, FormStateInterface $form_st
  * Duplicate of \Drupal\file\Plugin\Field\FieldWidget\FileWidget::submit(), with
  * a few changes:
  *   - Submit button is one level down compare to 'Upload' source's submit
- *     button.
+ *   button.
```
Comment (tienvx) [inline filefield_sources.module]: Please revert this change.

File: filefield_sources.module [hunk 6]
```diff
@@ -298,9 +302,9 @@ function filefield_sources_field_validate(&$element, FormStateInterface $form_st
...
  *   - Replace static in static::getWidgetState and static::setWidgetState by
- *     WidgetBase.
+ *   WidgetBase.
```
Comment (tienvx) [inline filefield_sources.module]: Please revert this change.

File: filefield_sources.module [hunk 7]
```diff
@@ -441,11 +445,13 @@ function filefield_sources_save_file($filepath, $validators = array(), $destinat
   $user = \Drupal::currentUser();
 
   // Begin building file object.
-  $file = entity_create('file', array(
-    'uri' => $filepath,
-    'uid' => $user->id(),
-    'status' => FILE_EXISTS_RENAME,
-  ));
+  $file = entity_create(
```
Comment (tienvx) [inline filefield_sources.module]: Please revert this change.

File: src/Plugin/FilefieldSource/Remote.php [hunk 8]
```diff
@@ -116,7 +116,7 @@ public static function value(array &$element, &$input, FormStateInterface $form_
         $pathinfo = pathinfo($filename);
       }
 
-      $filename = filefield_sources_clean_filename($filename, $field->settings['file_extensions']);
+      $filename = filefield_sources_clean_filename($filename, $field->settings('file_extensions'));
```
Comment (tienvx) [inline src/Plugin/FilefieldSource/Remote.php]: Please take a look at my notes in Attach.php

File: src/Plugin/FilefieldSource/Remote.php [hunk 9]
```diff
@@ -125,7 +125,7 @@ public static function value(array &$element, &$input, FormStateInterface $form_
       }
 
       // Perform basic extension check on the file before trying to transfer.
-      $extensions = $field->settings['file_extensions'];
+      $extensions = $field->settings('file_extensions');
```
Comment (tienvx) [inline src/Plugin/FilefieldSource/Remote.php]: Please take a look at my notes in Attach.php

File: src/Plugin/FilefieldSource/Attach.php [hunk 10]
```diff
@@ -154,23 +158,25 @@ public static function process(array &$element, FormStateInterface $form_state,
     }
 
     $ajax_settings = [
-      'url' => Url::fromRoute('file.ajax_upload'),
+      'callback' => [get_called_class(), 'uploadAjaxCallbacknew'],
```
Comment (tienvx) [inline src/Plugin/FilefieldSource/Attach.php]: Please replace:
```
[get_called_class(), 'uploadAjaxCallbacknew'],
```
by
```
$class = '\Drupal\file\Element\ManagedFile';
[$class, 'uploadAjaxCallback'],
```

File: src/Plugin/FilefieldSource/Attach.php [hunk 11]
```diff
@@ -346,4 +358,47 @@ public static function filePathValidate(array &$element, FormStateInterface $for
...
+    $form['#prefix'] .= $renderer->renderRoot($status_messages);
+    $output = $renderer->renderRoot($form);
+
+    $response = new AjaxResponse();
+    $response->setAttachments($form['#attached']);
+
+    return $response->addCommand(new ReplaceCommand(NULL, $output));
+  }
+
 }
```
Comment (tienvx) [inline src/Plugin/FilefieldSource/Attach.php]: Please remove these code, we don't need it any more.

File: src/Plugin/FilefieldSource/Attach.php [hunk 12]
```diff
@@ -14,6 +14,10 @@
 use Drupal\Core\Site\Settings;
 use Drupal\Core\Template\Attribute;
 use Drupal\Core\Url;
+use Drupal\Component\Utility\Html;
+use Symfony\Component\HttpFoundation\Request;
+use Drupal\Core\Ajax\AjaxResponse;
+use Drupal\Core\Ajax\ReplaceCommand;
```
Comment (tienvx) [inline src/Plugin/FilefieldSource/Attach.php]: Please remove these use statements, we don't need it any more.

File: src/Plugin/FilefieldSource/Attach.php [hunk 13]
```diff
@@ -154,23 +158,25 @@ public static function process(array &$element, FormStateInterface $form_state,
...
     ];
-
     $element['filefield_attach']['attach'] = [
       '#name' => implode('_', $element['#parents']) . '_attach',
       '#type' => 'submit',
       '#value' => t('Attach'),
       '#validate' => [],
-      '#submit' => ['filefield_sources_field_submit'],
+      '#submit' => ['file_managed_file_submit'],
       '#limit_validation_errors' => [$element['#parents']],
```
Comment (tienvx) [inline src/Plugin/FilefieldSource/Attach.php]: I think this submit callback is not valid in drupal core 8.0.x. Please revert this change.

File: src/Plugin/FilefieldSource/Attach.php [hunk 14]
```diff
@@ -154,23 +158,25 @@ public static function process(array &$element, FormStateInterface $form_state,
...
+      'callback' => [get_called_class(), 'uploadAjaxCallbacknew'],
       'options' => [
         'query' => [
           'element_parents' => implode('/', $element['#array_parents']),
-          'form_build_id' => $complete_form['form_build_id']['#value'],
         ],
       ],
-      'wrapper' => $element['#id'] . '-ajax-wrapper',
+      'wrapper' => $element['upload_button']['#ajax']['wrapper'],
       'effect' => 'fade',
```
Comment (tienvx) [inline src/Plugin/FilefieldSource/Attach.php]: Are there any reason to change this wrapper? Please take a look at this code in filefield_sources_field_process()
```
$new_wrapper = $field_element['#id'] . '-ajax-wrapper';
$element[$key][$subkey]['#ajax']['wrapper'] = $new_wrapper;
```

**Comment by tienvx** [comment]
Also, I suggest you can use some tools like ide (e.g. netbeans) and debugger (e.g. xdebug) to debug code easier. They can replace debug_print_backtrace and var_dump, or drupal's tools like dpm, kpr perfectly.

File: src/Plugin/FilefieldSource/Attach.php [hunk 15]
```diff
@@ -152,19 +152,21 @@ public static function process(array &$element, FormStateInterface $form_state,
...
     $ajax_settings = [
-      'url' => Url::fromRoute('file.ajax_upload'),
+      'callback' => [$class, 'uploadAjaxCallback'],
       'options' => [
         'query' => [
           'element_parents' => implode('/', $element['#array_parents']),
-          'form_build_id' => $complete_form['form_build_id']['#value'],
         ],
       ],
       'wrapper' => $element['#id'] . '-ajax-wrapper',
```
Comment (tienvx) [inline src/Plugin/FilefieldSource/Attach.php]: You were right. It was my my fault. Please change this line to:
```
'wrapper' => $element['upload_button']['#ajax']['wrapper'],
```

Please update this change to **other plugins** too.

File: src/Plugin/FilefieldSource/Attach.php [hunk 16]
```diff
@@ -152,19 +152,21 @@ public static function process(array &$element, FormStateInterface $form_state,
       );
       $element['filefield_attach']['#description'] = $description;
     }
-
+    $class = '\Drupal\file\Element\ManagedFile';
     $ajax_settings = [
-      'url' => Url::fromRoute('file.ajax_upload'),
+      'callback' => [$class, 'uploadAjaxCallback'],
```
Comment (tienvx) [inline src/Plugin/FilefieldSource/Attach.php]: Please update this change to **other plugins** too.

File: src/Plugin/FilefieldSource/Attach.php [hunk 17]
```diff
@@ -183,7 +185,11 @@ public static function process(array &$element, FormStateInterface $form_state,
    */
   public static function element($variables) {
     $element = $variables['element'];
-
+    $options = form_select_options($element['filename']);
+    $option_output = '';
+    foreach ($options as $key => $value) {
+      $option_output .= '<option value=' . $value["value"] . '>' . $value["label"] . '</option>';
+    }
```
Comment (tienvx) [inline src/Plugin/FilefieldSource/Attach.php]: Please move this code to the body of else condition bellow.
```
if (isset($attach_message)) {
  ...
}
else {
  // Move the code here.
}
```

**Comment by tienvx** [comment]
There are still 7 fails and 3 exceptions while running test cases. I will fix them latter after merging your code.

File: src/Plugin/FilefieldSource/Attach.php [hunk 18]
```diff
@@ -152,19 +152,21 @@ public static function process(array &$element, FormStateInterface $form_state,
...
     }
-
+    $class = '\Drupal\file\Element\ManagedFile';
     $ajax_settings = [
-      'url' => Url::fromRoute('file.ajax_upload'),
+      'callback' => [$class, 'uploadAjaxCallback'],
       'options' => [
         'query' => [
           'element_parents' => implode('/', $element['#array_parents']),
-          'form_build_id' => $complete_form['form_build_id']['#value'],
```
Comment (tienvx) [inline src/Plugin/FilefieldSource/Attach.php]: Do you have the reason to remove it? If not please keep it.
